### PR TITLE
RESTWS-823: Add REST resource for Alert Management

### PR DESCRIPTION
**Pull request by BartlomiejRasztabiga**
## Description of what I changed
Added REST resource for managing OpenMRS Alerts.

PR with documentation: [https://github.com/openmrs/openmrs-contrib-rest-api-docs/pull/143](https://github.com/openmrs/openmrs-contrib-rest-api-docs/pull/143)

## Issue I worked on
see https://issues.openmrs.org/browse/RESTWS-823

## Checklist: I completed these to help reviewers :)
- [X] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/AlertRecipientResource1_8.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+	@Override
+	public Model getGETModel(Representation rep) {
+		ModelImpl modelImpl = ((ModelImpl) super.getGETModel(rep))
+				.property("uuid", new StringProperty())
+				.property("display", new StringProperty());
+
+		if (rep instanceof DefaultRepresentation) {
+			modelImpl
+					.property(RECIPIENT, new RefProperty("#/definitions/UserGetRef"))
```
Comment (gitcliff) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/AlertRecipientResource1_8.java:106]: @BartlomiejRasztabiga  thanks for the work on this issue,
Could you fix the indenting here and in other places as done on line 62-69 above..

Comment (BartlomiejRasztabiga) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/AlertRecipientResource1_8.java]: @gitcliff Should be resolved now :)
I've kept all of them at one-tab indentation. I know that on GitHub it looks different than in IDE.

**Comment by BartlomiejRasztabiga** [comment]
@dkayiwa Can you see why the snyk job has failed? I can't access it

**Comment by dkayiwa** [comment]
That one you can ignore.

**Comment by BartlomiejRasztabiga** [comment]
@dkayiwa Have you got any change requests to this PR? :)

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertRecipientResource2_0.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+import org.openmrs.notification.Alert;
+import org.openmrs.notification.AlertRecipient;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Sub-resource for {@link AlertRecipient}
+ */
+@SubResource(parent = AlertResource2_0.class, path = "recipient", supportedClass = AlertRecipient.class, supportedOpenmrsVersions = {
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertRecipientResource2_0.java:44]: What does `recipient` mean?

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertRecipientResource2_0.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+	@Override
+	public DelegatingResourceDescription getRepresentationDescription(Representation rep) {
+		DelegatingResourceDescription description = new DelegatingResourceDescription();
+		description.addProperty("uuid");
+		description.addProperty("display");
+		description.addSelfLink();
+
+		if (rep instanceof DefaultRepresentation) {
+			description.addProperty(RECIPIENT, Representation.REF);
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertRecipientResource2_0.java:62]: I find these constants a bit different from what is else where.

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertController2_0Test.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+import org.openmrs.notification.AlertService;
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+
+import static org.junit.Assert.*;
```
Comment (dkayiwa) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertController2_0Test.java:33]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotusewildcardimports

**Comment by dkayiwa** [comment]
Did you add a controller test for each resource?

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertRecipientResource2_0.java (hunk 2 above)
Comment (BartlomiejRasztabiga) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertRecipientResource2_0.java:44]: It's a sub-resource of Alert resource. Each Alert has a set of AlertRecipients so this is the resource to access them.

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertRecipientResource2_0.java (hunk 3 above)
Comment (BartlomiejRasztabiga) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertRecipientResource2_0.java:62]: What do you mean? :)

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertController2_0Test.java (hunk 4 above)
Comment (BartlomiejRasztabiga) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertController2_0Test.java]: Should be resolved now

**Comment by BartlomiejRasztabiga** [comment]
> Did you add a controller test for each resource?

Right, I will add a controller test for `AlertRecipientResource2_0` as well :)

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertRecipientResource2_0.java (hunk 3 above)
Comment (dkayiwa) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertRecipientResource2_0.java:66]: Do existing resources also use constants for the same?

Comment (BartlomiejRasztabiga) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertRecipientResource2_0.java:66]: They don't but using constants for duplicate strings is a good practice.

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertRecipientController2_0Test.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+	@Test
+	public void shouldAddRecipientToAlert() throws Exception {
+		String json = "{\"recipient\": \"" + RestTestConstants1_8.USER_UUID + "\"}";
+		handle(newPostRequest("alert/" + EMPTY_ALERT_UUID + "/recipient", json));
+
+		Alert existingAlert = getAlertByUuid(EMPTY_ALERT_UUID);
+		assertNotNull(existingAlert);
+
+		assertEquals(1, existingAlert.getRecipients().size());
```
Comment (dkayiwa) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertRecipientController2_0Test.java:81]: How do i tell that the recipients size was not 1 before the above?

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertRecipientController2_0Test.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	@Test
+	public void shouldPurgeRecipientFromAlert() throws Exception {
+		handle(newDeleteRequest(getURI() + "/" + getUuid()));
+
+		Alert existingAlert = getAlertByUuid(NEW_ALERT_UUID);
+		assertNotNull(existingAlert);
+
+		assertEquals(0, existingAlert.getRecipients().size());
```
Comment (dkayiwa) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertRecipientController2_0Test.java:91]: How do i tell that the size was originally greater than 0?

**Comment by dkayiwa** [comment]
How did you deal with Roles on the alert management screen?

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertRecipientController2_0Test.java (hunk 5 above)
Comment (BartlomiejRasztabiga) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertRecipientController2_0Test.java:86]: Resolved here: https://github.com/openmrs/openmrs-module-webservices.rest/pull/485/commits/557551160367c70d0408f92cd260475ac7931f0b
`alert.getRecipients()` returns null if there are no recipients

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertRecipientController2_0Test.java (hunk 6 above)
Comment (BartlomiejRasztabiga) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AlertRecipientController2_0Test.java:101]: Resolved here: https://github.com/openmrs/openmrs-module-webservices.rest/pull/485/commits/be2f9dfa11de9a3bd57d4bc14b9f0a0d49e59c27

**Comment by BartlomiejRasztabiga** [comment]
> How did you deal with Roles on the alert management screen?

@dkayiwa I though we wouldn't need that feature on the REST API side. 
What Roles section does, is that adds each member of a given Role as alert's recipient. So presumably, REST API client could beforehand gather all users from those roles and pass a list of gathered users when POSTing new Alert.

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertResource2_0.java [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+	public Alert newDelegate() {
+		Alert alert = new Alert();
+		alert.setRecipients(new HashSet<>()); // to prevent NullPointerException
+		return alert;
+	}
+
+	@Override
+	public Alert getByUniqueId(String uniqueId) {
+		boolean includeRetired = true;
+		List<Alert> alerts = Context.getAlertService().getAllAlerts(includeRetired);
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertResource2_0.java:225]: This does not look efficient.

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertResource2_0.java [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+	}
+
+	@Override
+	public Alert save(Alert alert) {
+		return Context.getAlertService().saveAlert(alert);
+	}
+
+	@Override
+	public void purge(Alert alert, RequestContext context) throws ResponseException {
+		if (alert == null) {
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertResource2_0.java:241]: How would one call purge with a null alert?

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertResource2_0.java (hunk 7 above)
Comment (BartlomiejRasztabiga) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertResource2_0.java:225]: There is no method to `getAlertByUuid` in `AlertService`. I could add one but then we would have to move this resource to omod-2.5, after releasing the new AlertService for OpenMRS 2.5.

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertResource2_0.java (hunk 8 above)
Comment (BartlomiejRasztabiga) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertResource2_0.java:241]: I'm not sure but I've seen other Resources to include such a null check, for example: `ConceptResource1_8, CohortResource1_8, DrugResource1_8` etc.

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertResource2_0.java [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+		return description;
+	}
+
+	@PropertyGetter("display")
+	public String getDisplayProperty(Alert alert) {
+		return alert.getText();
+	}
+
+	@PropertySetter(RECIPIENTS)
+	public static void setRecipientsProperty(Alert alert, Set<AlertRecipient> newRecipients)
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertResource2_0.java:108]: If these are new recipients, shouldn't we just add them to the alert?

Comment (BartlomiejRasztabiga) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AlertResource2_0.java:108]: These may be also updated alerts. For example, if previously we've had recipient ABC and now we're updating it with BCD then I will get hibernate error on save, that in this session there are more than 1 copies of BC objects.

**Pull request merged**